src/kudu/consensus/leader_election.h

```diff
@@ -97,37 +122,131 @@ class VoteCounter {
+
+  // Iterates over all the votes that have come in so far and collates them
+  // corresponding to each UUID term pair for the purpose of determining if
+  // one of those UUIDs could have won an election after the `term` provided
+  // when the last known leader's region was `leader_region`. It also computes
+  // the min_term greater than `term` in which one of the UUIDs from the
+  // `leader_region` had voted. This function only considers votes that are
+  // relevant as per the leader election quorum requirement.
+  void ConstructRegionWiseVoteCollation(
+      const int64_t term,
+      const std::string& leader_region,
+      VoteHistoryCollation* vote_collation,
+      VoteHistoryIteratorMap* it_map,
+      int64_t* min_term) const;
+
+  // Updates the region-wise vote collation represented by `vote_collation`
+  // to consider votes only after `term` provided. It also advances the
+  // iterators in `it_map` such that they represent a point in voting history
+  // after `term`. It also computes the min_term greater than `term` in which
+  // one of the UUIDs from the `leader_region` had voted.
+  void UpdateRegionWiseVoteCollation(
+      const int64_t term,
+      const std::string& leader_region,
+      VoteHistoryCollation* vote_collation,
+      VoteHistoryIteratorMap* it_map,
+      int64_t* min_term) const;
+
+  // Optimizer function which tries to recursively figure out the next leader
+  // region since the last known leader. It accepts two parameters, term and
+  // region of the last known leader. It returns the UUID and term of a node
+  // that could potentially be the leader in a subsequent term that the
+  // candidate might not be aware of. It is possible that the next possible
+  // leader cannot be determined in which case, the situation is reflected in
+  // the status.
+  PotentialNextLeaderResponse GetPotentialNextLeader(
+      const int64_t term, const std::string& leader_region) const;
+
   // Returns a couple of booleans - the first denotes if the election quorum
   // has been satisfied and the second denotes if it can still be satisfied.
   std::pair<bool, bool> GetQuorumState() const;
 
   // Fetches topology information required by the flexible vote counter.
-  void FetchTopologyInfo(
-      std::map<std::string, int>* voter_distribution,
-      std::map<std::string, int>* le_quorum_requirement);
+  void FetchTopologyInfo();
+
+  // Computes a pessimistic leader election quorum.
+  void ComputePessimisticLeaderElectionQuorum();
+
+  // Generic log prefix.
+  std::string LogPrefix() const;
+
+  // Term of this election.
+  const int64_t election_term_;
 
   // Mapping from each region to number of active voters.
   std::map<std::string, int> voter_distribution_;
 
   // Mapping from each region to number of yes votes required.
   std::map<std::string, int> le_quorum_requirement_;
 
+  // Worst case quorum requirement for leader election.
+  std::map<std::string, int> pessimistic_le_quorum_;
```

Comment:
nit: how about `pessimistic_election_quorum_` and avoid the abbreviation? (same with the above var; how about s/le/election/)